pmd-core/src/test/java/net/sourceforge/pmd/cache/FileAnalysisCacheTest.java

```diff
@@ -120,7 +123,7 @@ void testStorePersistsFilesWithViolations() throws IOException {
         when(rule.getLanguage()).thenReturn(mock(Language.class));
         when(rv.getRule()).thenReturn(rule);
 
-        cache.startFileAnalysis(sourceFile).onRuleViolation(rv);
```

Comment:
this worked because the file listener is created AFTER the call to `isUpToDate`, so `updatedResultsCache` already contained an entry for this file, otherwise a local instance was created, and when `isUpToDate` is eventually called, that remained pristine.

The new implementation is completely agnostic of the order of invocation between `startFileAnalysis` and `isUpToDate`, only requires that `isUpToDate` is actually checked before reporting violations (which makes sense, as this is a cache)